mypy/constraints.py

```diff
@@ -339,6 +340,26 @@ def _infer_constraints(
     if isinstance(actual, AnyType) and actual.type_of_any == TypeOfAny.suggestion_engine:
         return []
 
+    # type[A | B] is always represented as type[A] | type[B] internally.
+    # This makes our constraint solver choke on type[T] <: type[A] | type[B],
+    # solving T as generic meet(A, B) which is often `object`. Force unwrap such unions
+    # if both sides are type[...] or unions thereof. See `testTypeVarType` test
+    def _is_type_type(tp: ProperType) -> TypeGuard[TypeType | UnionType]:
```

Comment:
Currently nested functions add some overhead when compiled with mypyc (and also when not compiled I guess), and this is a fairly hot function. Can you change the functions to be module-level functions instead?